faunadb-scala/src/main/scala/faunadb/FaunaClient.scala

```diff
@@ -248,50 +240,89 @@ class FaunaClient private (connection: Connection) {
   def syncLastTxnTime(timestamp: Long): Unit =
     connection.syncLastTxnTime(timestamp)
 
-  private def handleNetworkExceptions[A]: PartialFunction[Throwable, Future[A]] = {
-    case ex: ConnectException => Future.failed(new UnavailableException(ex.getMessage, ex))
-    case ex: TimeoutException => Future.failed(new TimeoutException(ex.getMessage))
+  private def parseResponseBody(response: FullHttpResponse)(implicit ec: ExecutionContext): Future[JsonNode] = {
+    def parse: Future[Option[JsonNode]] = Future(Option(json.readTree(new ByteBufInputStream(response.content()))))
+
+    parse.flatMap {
+      case Some(json) => Future.successful(json)
+      case None => Future.failed(new IOException("Invalid JSON."))
+    }
+  }
+
+  private def handleSuccessResponse(response: FullHttpResponse)(implicit ec: ExecutionContext): Future[Value] = {
+    def getResource(body: JsonNode): Future[JsonNode] = Option(body.get("resource")) match {
+      case Some(resource) => Future.successful(resource)
+      case None => Future.failed(new IOException("Invalid JSON."))
+    }
+
+    def parseValue(resource: JsonNode): Future[Value] = resource match {
+      case _: NullNode => Future.successful(NullV)
+      case _: JsonNode => Future(json.treeToValue[Value](resource, classOf[Value]))
+    }
+
+    val result: Future[Value] =
+      for {
+        body <- parseResponseBody(response)
+        resource <- getResource(body)
+        value <- parseValue(resource)
+      } yield value
+
+    result.andThen {
+      case _ => response.release()
+    }
   }
 
   private def handleErrorResponse(response: FullHttpResponse)(implicit ec: ExecutionContext): Future[Nothing] = {
-    def parseError(): Future[QueryErrorResponse] = Future {
+    def parseErrors(): Future[QueryErrorResponse] = {
       val statusCode = response.status().code()
-      val errors = Option(parseResponseBody(response).get("errors").asInstanceOf[ArrayNode])
-      val parsedErrors = errors.map(_.iterator().asScala.map(json.treeToValue(_, classOf[QueryError])).toIndexedSeq).getOrElse(IndexedSeq.empty)
-      val error = QueryErrorResponse(statusCode, parsedErrors)
-      error
-    }.recoverWith {
-      case e: FaunaException => Future.failed(e)
-      case unavailable if response.status().code() == 503 => Future.failed(new UnavailableException("Service Unavailable: Unparseable response.", unavailable))
-      case unknown => Future.failed(new UnknownException(s"Unparseable service $unknown response.", unknown))
-    }.andThen {
-      case _ => response.release()
+
+      def getErrors(body: JsonNode): Future[Iterator[JsonNode]] = Option(body.get("errors")) match {
+        case Some(errors) => Future(errors.asInstanceOf[ArrayNode].iterator().asScala)
```

Comment:
any chance of this `errors.asInstanceOf[ArrayNode].iterator().asScala` throw an exception? otherwise you can use `Future.successful()`

Reply:
This is a good point! If the server responds with the proper respond, it should not. But we cannot guarantee that at this point, and if the value cannot be casted into `ArrayNode` it will throw an error!

Reply:
We could try the following:

```
def getErrors(body: JsonNode): Try[Iterator[JsonNode]] = Option(body.get("errors")) match {
  case Some(errors) => Try(errors.asInstanceOf[ArrayNode].iterator().asScala)
  case None => Success(Iterator.empty)
}
```

and then:

```
val result: Future[QueryErrorResponse] =
  for {
    body <- parseResponseBody(response)
    errors <- Future.fromTry(getErrors(body))
    queryErrors <- parseErrors(errors)
  } yield QueryErrorResponse(statusCode, queryErrors)
```

We should ponder clarity vs performance.